### PR TITLE
types(collection): reduce* method signatures

### PR DESCRIPTION
**Pull request by Renegade334**
**Please describe the changes this PR makes and why it should be merged:**

Small changes relating to the typed method definitions for `reduce()` and `reduceRight()`:
- Fix the "without-initial-value" signature of `reduceRight()`.
  - When `initialValue` was not provided, this method would previously accept a callback with any arbitrary type for `previousValue`, or infer the type as `unknown` without explicit type annotation:
    ```ts
    declare const coll: Collection<string, number>;
    coll.reduceRight((a, x) => a + x); // error: 'a' is of type 'unknown'
    coll.reduceRight((a: string, x) => a.repeat(x)); // compiles OK, but will raise a TypeError at runtime
    ```
  - With this change, if an initial value for the accumulator is not provided, the callback's `previousValue` type is always the collection value type, and will raise an error otherwise.
- Separate out "accumulator type = collection value type" and "accumulator type = arbitrary type" overload signatures, à la TypeScript's library definitions for `Array.prototype.reduce`.
  - This changes the type behaviour in one specific set of cases, which is where a value for `initialValue` is provided with a type that is assignable to, but not equal to, the collection value type:
    ```ts
    declare const coll: Collection<string, number>;
    declare const initial: 1 | 2 | 3;
    
    coll.reduce((a, x) => a + x, initial); // Type 'number' is not assignable to type '1 | 2 | 3'
    ```
  - Previously, `InitialValue` would be inferred as `1 | 2 | 3`, leading to compilation errors.
  - With this change, if `initialValue` is assignable to the collection value type, the method signature will preferentially use the collection value type for the callback, rather than narrowing to the type of `initialValue`. This matches the behaviour of `Array.prototype.reduce` in TypeScript.

Also adds one missing `reduceRight()` test (reduce empty collection with initial value).

### CODE REVIEW

File: packages/collection/__tests__/collection.test.ts [hunk 1]
```diff
@@ -705,12 +705,12 @@ describe('reduce() tests', () => {
...
 	test('reduce collection into a single value with initial value', () => {
 		const sum = coll.reduce((a, x) => a + x, 0);
-		expect<number>(sum).toStrictEqual(6);
+		expect(sum).toStrictEqual(6);
 	});
 
 	test('reduce collection into a single value without initial value', () => {
 		const sum = coll.reduce((a, x) => a + x);
-		expect<number>(sum).toStrictEqual(6);
+		expect(sum).toStrictEqual(6);
```
[the remark this answers is not in the document]
Comment (kyranet) [reply packages/collection/__tests__/collection.test.ts:713]: That's true, and I agree it kind of shows as a bit of inconsistency. We use the type parameter to determine whether or not the type of an expression (in this case, the return type of a method) is correct.

This added benefit is why formatters' [tests](https://github.com/discordjs/discord.js/blob/main/packages/formatters/__tests__/formatters.test.ts) use it extensively. I believe it has a place in `collection` as well so not only it's runtime-tested, but also type-tested.

[the remark this answers is not in the document]
Comment (Renegade334) [reply packages/collection/__tests__/collection.test.ts:713]: In which case, I'll add them in for the reduce\* tests. It does make sense to be explicit here, given that those methods have variable return types based on how the method's called.

[the remark this answers is not in the document]
Comment (kyranet) [reply packages/collection/__tests__/collection.test.ts:713]: That is fine to me, thanks :)

It makes sense in other methods as well, such as `find*` and `filter*`, to ensure the types are correctly handled, but that's outside this PR's scope (feel free to open a PR to update those)

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by didinele — APPROVED** [review]
